cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/submit/AbstractSparkSubmitter.java

```diff
@@ -216,14 +218,61 @@ protected void submit(SparkRuntimeContext runtimeContext,
 
   private abstract static class SparkJobFuture<V> extends AbstractFuture<V> {
 
+    private static final Logger LOG = LoggerFactory.getLogger(SparkJobFuture.class);
+    private final AtomicBoolean done;
+    private final SparkRuntimeContext context;
+
+    protected SparkJobFuture(SparkRuntimeContext context) {
+      this.done = new AtomicBoolean();
+      this.context = context;
+    }
+
     @Override
-    public boolean set(@Nullable V value) {
-      return super.set(value);
+    protected boolean set(V value) {
+      if (done.compareAndSet(false, true)) {
+        return super.set(value);
+      }
+      return false;
     }
 
     @Override
-    public boolean setException(Throwable throwable) {
-      return super.setException(throwable);
+    protected boolean setException(Throwable throwable) {
+      if (done.compareAndSet(false, true)) {
+        return super.setException(throwable);
+      }
+      return false;
+    }
+
+    @Override
+    public boolean cancel(boolean mayInterruptIfRunning) {
+      if (!done.compareAndSet(false, true)) {
+        return false;
+      }
+
+      try {
+        cancelTask();
+        return super.cancel(mayInterruptIfRunning);
+      } catch (Throwable t) {
+        // Only log and reset state, but not propagate since Future.cancel() doesn't expect exception to be thrown.
+        LOG.warn("Failed to cancel Spark execution for {}.", context, t);
+        done.set(false);
+        return false;
+      }
+    }
+
+
+    @Override
+    protected final void interruptTask() {
+      // Final it so that it cannot be overridden. This method gets call after the Future state changed
+      // to cancel, hence cannot have the caller block until cancellation is done.
```

Comment:
I think it would help to add a javadoc to this class: its prupose it to introduce a new "lifecycle" method interruptTask, that should be used instead of cancelTask, and the reason why.